apps/tool/tool.1.md
% SPOC\_TOOL(1) SPOC User's Manual | Version 0.1
% spoc@spocfile.xyz
% December 25, 2021

# NAME

spoc\_tool - SPOC file tool

# USAGE

spoc\_tool [*options*] [*input_filename*] [*output_filename*]

# DESCRIPTION

Collection of tools to run on a SPOC file

# OPTIONS

\-\-help, -h
:   Get help

\-\-verbose, -v
:   Set verbose mode ON

\-\-version, -e
:   Print version information and exit

\-\-random-seed=*#*, -r *#*
:   Set the random seed equal to *#*

\-\-resolution=*#*
:   Set the resolution used for upsampling

\-\-field-filename=*FN*
:   Set the filename of the text file used with the *set-field* command.
    This file (or named pipe) will contain newline separated values for
    the field being set.

# COMMANDS

\-\-get-field=*F*, -g *F*
:   Get point field *F* and write it to stdout as text. *F* can be one of
    'x', 'y', 'z', 'c', 'p', 'i', 'r', 'g', 'b', or 'e#', where the '#'
    after the 'e' specifies the extra field number.

\-\-recenter-xy
:   Recenter the point cloud by subtracting the mean X and Y value from
    each point's X and Y value

\-\-recenter-xyz
:   Same as recenter-xy, but also recenter the Z value

\-\-resize-extra=*#*
:   Change the number of extra fields in each point record to *#*

\-\-set-field=*F*, -s *F*
:   Set point field *F* by reading it from a file specified by the
    --field-filename option. *F* can be one of 'x', 'y', 'z', 'c', 'p',
    'i', 'r', 'g', 'b', or 'e#', where the '#' after the 'e' specifies
    the extra field number.

\-\-subtract-min-xy
:   Subtract minimum X and Y values from all X and Y values, thereby
    translating the point cloud's minimum to 0.0, 0.0.

\-\-subtract-min-xyz
:   Same as subtract-min-xy, but also subtract min from the Z value

\-\-upsample-classifications=*FN*
:   Upsample the classifications field from a low resolution SPOC file
    into the high resolution SPOC file specified by *FN* and write the
    result to the output file. The SPOC file being upsampled must have
    first been subsampled from *FN* using the *spoc_filter* application.
    The resolution should also be set using the *resolution* option, and
    this resolution should be the same resolution used for subsampling.

# SEE ALSO

SPOC\_FILTER(1)
SPOC\_TRANSFORM(1)
SPOC\_INFO(1)


apps/tool/tool.cpp
#include "spoc/app_utils.h"
#include "spoc/spoc.h"
#include "tool.h"
#include "tool_cmd.h"
#include <cctype>
#include <iostream>
#include <stdexcept>

int main (int argc, char **argv)
{
    using namespace std;
    using namespace spoc::app_utils;
    using namespace spoc::file;
    using namespace spoc::io;
    using namespace spoc::tool_app;
    using namespace spoc::tool_cmd;

    try
    {
        // Parse command line
        const args args = get_args (argc, argv,
                string (argv[0]) + " [options] [input] [output]");

        // If version was requested, print it and exit
        if (args.version)
        {
            cout << "Version "
                << static_cast<int> (spoc::MAJOR_VERSION)
                << "."
                << static_cast<int> (spoc::MINOR_VERSION)
                << endl;
            return 0;
        }

        // If you are getting help, exit without an error
        if (args.help)
            return 0;

        // Check the arguments
        if (args.command.name.empty ())
            throw runtime_error ("No command was specified");

        // Show args
        if (args.verbose)
        {
            clog << "verbose\t" << args.verbose << endl;
            clog << "field_fn\t'" << args.field_fn << "'" << endl;
            clog << "input_fn\t'" << args.input_fn << "'" << endl;
            clog << "output_fn\t'" << args.output_fn << "'" << endl;
            clog << "resolution\t'" << args.resolution << "'" << endl;
            clog << "command: " << args.command.name << "=" << args.command.params << endl;
        }

        // Get the input stream
        input_stream is (args.verbose, args.input_fn);

        // Read the input file
        spoc_file f = read_spoc_file (is ());

        if (args.command.name == "get-field")
        {
            string s = args.command.params;
            const auto l = consume_field_name (s);
            get_field (f, cout, l);

            // Short-circuit so that the SPOC file is not written
            return 0;
        }
        else if (args.command.name == "recenter-xy")
            f = recenter (f);
        else if (args.command.name == "recenter-xyz")
            f = recenter (f, true);
        else if (args.command.name == "resize-extra")
        {
            const int new_size = std::stoi (args.command.params);
            if (new_size < 0)
                throw runtime_error ("The number of extra fields must be >= 0");
            if (new_size > 255)
                throw runtime_error ("The number of extra fields must be < 256");
            f = resize_extra (f, new_size);
        }
        else if (args.command.name == "set-field")
        {
            if (args.field_fn.empty ())
                throw runtime_error ("You must set the 'field-filename' option when using the 'set-field' command");
            if (args.verbose)
                clog << "Opening " << args.field_fn << " for reading" << endl;
            ifstream field_ifs (args.field_fn);
            if (!field_ifs)
                throw runtime_error ("Could not open file for reading");
            string s = args.command.params;
            const auto l = consume_field_name (s);
            f = set_field (f, field_ifs, l);
        }
        else if (args.command.name == "subtract-min-xy")
            f = subtract_min (f);
        else if (args.command.name == "subtract-min-xyz")
            f = subtract_min (f, true);
        else if (args.command.name == "upsample-classifications")
        {
            // Check the resolution parameter
            if (args.resolution <= 0.0)
                throw runtime_error ("The 'resolution' option must be set to a value > 0.0 when using the 'upsample' command");

            // Open the file
            const auto fn = args.command.params;
            if (args.verbose)
                clog << "Opening " << fn << " for reading" << endl;
            ifstream ifs (fn);
            if (!ifs)
                throw runtime_error ("Could not open file for reading");
            // Read the spoc file
            const auto g = read_spoc_file (ifs);
            f = upsample_classifications (f, g, args.resolution, args.verbose, std::clog);
        }
        else
            throw runtime_error ("An unknown command was encountered");

        // Get the output stream
        output_stream os (args.verbose, args.output_fn);

        // Write it out
        write_spoc_file (os (), f);

        return 0;
    }
    catch (const exception &e)
    {
        cerr << e.what () << endl;
        return -1;
    }
}
